### PR TITLE
Computing groupwise metrics 

### PR DESCRIPTION
**Pull request by lastmansleeping**
@jakemannix Should be a simple, but very useful, addition. Creating this PR against the already existing branch since I'm using the latest code to avoid merge conflicts.

### CODE REVIEW

File: python/ml4ir/model/metrics/metrics_helper.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    old_rank_col: str,
+    new_rank_col: str,
+    group_keys: List[str] = [],
+):
+    """Compute metrics on a data batch/chunk"""
+
+    # Select clicked records
+    df_clicked = df[df[label_col] == 1.0]
+
+    if group_keys:
```
Comment (jakemannix) [inline python/ml4ir/model/metrics/metrics_helper.py]: does this check if `group_keys` is empty?

File: python/ml4ir/model/ranking_model.py [hunk 2]
```diff
@@ -245,24 +247,17 @@ def predict(
...
-                    )
-                    predictions_dict[feature_node_name] = tf.strings.regex_replace(
-                        str_feature, "\x00", ""
-                    )
-
-            predictions_df = pd.DataFrame(predictions_dict)
+            predictions_df = self._convert_predictions_to_df(
+                predictions_dict, self.feature_config.get_features_to_log()
+            )
             if logs_dir:
```
Comment (jakemannix) [inline python/ml4ir/model/ranking_model.py]: collapse these two if's together?  `if logs_dir and os.path.isfile(outfile):`

File: python/ml4ir/model/ranking_model.py [hunk 3]
```diff
@@ -286,15 +281,93 @@ def predict(
...
+    def _get_predict_fn(self, inference_signature, features_to_return):
         if self.is_compiled:
             infer = self.model
         else:
             # If SavedModel was loaded without compilation
             infer = self.model.signatures[inference_signature]
 
         # Get features to log
-        features_to_log = self.feature_config.get_features_to_log(key="node_name")
+        features_to_log = [f.get("node_name", f["name"]) for f in features_to_return]
```
Comment (jakemannix) [inline python/ml4ir/model/ranking_model.py]: `features_to_log = [f.get(... )] + ["new_score", "new_rank"]` ?

File: python/ml4ir/model/ranking_model.py [hunk 4]
```diff
@@ -351,32 +424,20 @@ def _predict_score(features, label):
...
+        # If feature is a string, convert back from bytes to string
+        for feature_info in features:
+            feature_node_name = feature_info.get("node_name", feature_info["name"])
+            if feature_info["feature_layer_info"]["type"] == FeatureTypeKey.STRING:
+                str_feature = tf.strings.unicode_encode(
+                    tf.cast(predictions_dict[feature_node_name], tf.int32),
+                    output_encoding="UTF-8",
+                )
+                predictions_dict[feature_node_name] = tf.strings.regex_replace(
+                    str_feature, "\x00", ""
```
Comment (jakemannix) [inline python/ml4ir/model/ranking_model.py]: what is `\x00` in python?

Comment (lastmansleeping) [reply python/ml4ir/model/ranking_model.py]: Actually I am not sure what this is. This is the default padding character on tensorflow.

Comment (lastmansleeping) [reply python/ml4ir/model/ranking_model.py]: I should leave a TODO for the future to fix this. Nothing critical though.

File: python/ml4ir/model/metrics/metrics_helper.py (hunk 1 above)
Comment (lastmansleeping) [reply python/ml4ir/model/metrics/metrics_helper.py]: Yep.

**Comment by jakemannix** [comment]
> Ok, ran locally, lgtm:
> 
> ```
> INFO: 2020-03-11 14:36:24.877 
> Final model saved to : models/test/final
> INFO: 2020-03-11 14:36:26.527 
> Overall Metrics: 
> old_ACR               1.666903
> new_ACR               1.939631
> old_MRR               0.780660
> new_MRR               0.698845
> perc_improv_ACR     -16.361312
> perc_improv_MRR     -10.480279
> query_count        1408.000000
> dtype: float64
> INFO: 2020-03-11 14:36:27.505 
> Model predictions written to -> logs/test/model_predictions.csv
> INFO: 2020-03-11 14:36:27.505 
> Done! Elapsed time: 00:00:17
> ```

of course, ideally we have dummy/test data which *improves* with training ;P

**Comment by lastmansleeping** [comment]
Note to self
- add dummy groups with test data
- log group keys

**Pull request merged**